x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/methods/import_rule.ts

```diff
@@ -35,29 +35,34 @@ export const importRule = async ({
   prebuiltRuleAssetClient,
   mlAuthz,
 }: ImportRuleOptions): Promise<RuleResponse> => {
-  const { ruleToImport, overwriteRules, allowMissingConnectorSecrets } = importRulePayload;
+  const { ruleToImport, overwriteRules, overrideFields, allowMissingConnectorSecrets } =
+    importRulePayload;
+  // For backwards compatibility, immutable is false by default
+  const rule = { ...ruleToImport, immutable: false, ...overrideFields };
 
   await validateMlAuth(mlAuthz, ruleToImport.type);
 
   const existingRule = await getRuleByRuleId({
     rulesClient,
-    ruleId: ruleToImport.rule_id,
+    ruleId: rule.rule_id,
   });
 
   if (existingRule && !overwriteRules) {
-    throw createBulkErrorObject({
+    throw createRuleImportErrorObject({
       ruleId: existingRule.rule_id,
-      statusCode: 409,
+      type: 'conflict',
       message: `rule_id: "${existingRule.rule_id}" already exists`,
     });
   }
 
   if (existingRule && overwriteRules) {
-    const ruleWithUpdates = await applyRuleUpdate({
+    let ruleWithUpdates = await applyRuleUpdate({
       prebuiltRuleAssetClient,
       existingRule,
-      ruleUpdate: ruleToImport,
+      ruleUpdate: rule,
     });
+    // applyRuleUpdate prefers the existing rule's values for `rule_source` and `immutable`, but we want to use the importing rule's calculated values
+    ruleWithUpdates = { ...ruleWithUpdates, ...overrideFields };
```

Comment:
`applyRuleUpdate` shouldn't call `await calculateRuleSource` anymore here if `overrideFields` were passed - we've already calculated rule source and immutable. The current code will call `await prebuiltRuleAssetClient.fetchAssetsByVersion` 1000 times for 1000 prebuilt rules being imported.

Probably it would be better to refactor `applyRuleUpdate` and move the `await calculateRuleSource` call outside of it.

Same for `applyRulePatch`.

Reply:
@xcrzx Please review this.

Reply:
@banderror yeah, agreed this definitely needs to be addressed. This is related to our [other conversation](https://github.com/elastic/kibana/pull/190198#discussion_r1788279453), as well. 

I'm going to spend some time tomorrow trying to refactor these various functions into something more manageable, but in the meantime I have a broad question for the team:

> In what situations do we want to derive `rule_source` from `immutable` (as we currently do in `calculateRuleSource`, and in what situations should we calculate it from available assets?

Reply:
I agree with what @banderror suggests. We are calculating the rule source here, only to immediately overwrite it with the values from `overrideFields`. From a performance standpoint, it would be better to skip the rule source calculation altogether if we already have values in `overrideFields`.